crates/polars-plan/src/logical_plan/optimizer/predicate_pushdown/utils.rs

```diff
@@ -312,27 +237,24 @@ where
             .to_field(&input_schema, Context::Default, expr_arena)
             .unwrap();
 
-        // we check if predicates can be done on the input above
-        // this can only be done if the current projection is not a projection boundary
-        let is_boundary = projection_is_definite_pushdown_boundary(*projection_node, expr_arena);
+        // should have been handled earlier by `pushdown_and_continue`.
+        assert_aexpr_allows_predicate_pushdown(*projection_node, expr_arena);
```

Comment:
Can we only call this an compile the function when `debug_assertions` e.g. in debug mode. I don't want every node to call this as it does allocations and traversals and there are users with a lot of column projections.